journalpump/senders/websocket.py

```diff
@@ -0,0 +1,286 @@
+from aiohttp_socks.utils import Proxy
+from concurrent.futures import TimeoutError as ConnectionTimeoutError
+from janus import Queue
+from threading import Thread
+from urllib.parse import urlparse
+
+import asyncio
+import logging
+import random
+import socket
+import ssl
+import time
+import websockets
+
+
+class ExponentialBackoff:
+    """
+    An implementation of the exponential backoff algorithm with full jitter described at
+    https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+    """
+
+    def __init__(self, base: float, factor: float, maximum: float, jitter: bool):
+        if base <= 0:
+            raise ValueError("base must be positive")
+        if factor <= 0:
+            raise ValueError("factor must be positive")
+        if maximum < base:
+            raise ValueError("maximum must be greater than or equal to base")
+        self._base = base
+        self._factor = factor
+        self._maximum = maximum
+        self._jitter = jitter
+        self._attempts = 0
+
+    def next_sleep(self) -> float:
+        result = self._base * (self._factor ** self._attempts)
+        if result <= self._maximum:
+            self._attempts += 1
+        else:
+            result = self._maximum
+
+        if self._jitter:
+            return random.uniform(0, result)
+
+        return result
+
+    def reset(self) -> None:
+        self._attempts = 0
+
+
+class WebsocketRunner(Thread):
+    def __init__(self, *, websocket_uri, socks5_proxy_url, ssl_enabled, ssl_ca, ssl_key, ssl_cert):
+        super().__init__()
+        self.setDaemon(True)
+        self.log = logging.getLogger(self.__class__.__name__)
+        self.websocket_uri = websocket_uri
+        self.socks5_proxy_url = socks5_proxy_url
+        self.ssl_enabled = ssl_enabled
+        self.ssl_ca = ssl_ca
+        self.ssl_key = ssl_key
+        self.ssl_cert = ssl_cert
+        self.websocket_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.websocket_loop)
+        self.stop_event = asyncio.Event()
+        self.queue = None
+        self.running = True
+        self.connected = False
+        self.backoff = ExponentialBackoff(
+            base=10, factor=1.8, maximum=60, jitter=True
+        )
+        # prevent websockets from logging message contents when we're otherwise in DEBUG mode
+        logging.getLogger("websockets").setLevel(logging.INFO)
+
+        self.socks5_proxy = None
+        if self.socks5_proxy_url:
+            self.socks5_proxy = Proxy.from_url(self.socks5_proxy_url, loop=self.websocket_loop)
+
+    async def consumer_handler(self, websocket):
+        # Dummy consumer to read and ignore messages from the websocket
+        while self.running:
+            _ = await websocket.recv()
+
+    async def producer_handler(self, websocket):
+        while self.running:
+            messages = await self.queue.async_q.get()
+
+            for message in messages:
```

Comment:
Going forward probably makes sense to add batch support so that these can all be sent in one request. Possibly also some configurable minimum send interval so that there won't be 1000 requests per second if there are 1000 evenly timed log messages per second.

Reply:
Yes indeed, the idea is to batch individual log messages and do end-to-end compression from here to kafkapump in a way that nodecom does not need to decompress and open up the batches. But that shall happen in the next iteration, at this point I'll just make the basic thing work.